unixfs/hamt/fetcher.go

```diff
@@ -0,0 +1,321 @@
+				f.hits++
+				f.mainLoopSendResult(j)
+				continue
+			}
+			if j.idx != -1 {
+				f.misses++
+				// job is not currently running so
+				// move job to todoFirst so that it will be done on the
+				// next batch job
+				f.todo.remove(j)
+				f.todoFirst = j
+			} else {
+				// job already running
+				f.nearMisses++
+			}
+			want = id
+		case bj := <-f.done:
+			f.doneCnt += len(bj.jobs)
+			f.cidCnt += len(bj.cids)
+			f.launch()
+			if want != nil {
+				j := f.jobs[want]
+				if j.res.vals != nil {
+					f.mainLoopSendResult(j)
+					want = nil
+				}
+			}
+			log.Infof("fetcher: batch job done")
+			f.mainLoopLogStats()
+		case <-f.ctx.Done():
+			if !f.idle {
+				// wait unit batch job finishes
+				<-f.done
+			}
+			log.Infof("fetcher: exiting")
+			f.mainLoopLogStats()
+			return
+		}
+	}
+}
+
+func (f *fetcher) mainLoopAddJob(hamt *Shard) (*job, error) {
+	children, err := hamt.missingChildShards()
+	if err != nil {
+		return nil, err
+	}
+	if len(children) == 0 {
+		return nil, nil
+	}
+	j := &job{id: hamt, cids: children}
+	if len(j.cids) > batchSize {
+		// programmer error
+		panic("job size larger than batchSize")
+	}
+	f.todo.push(j)
+	f.jobs[j.id] = j
+	return j, nil
+}
+
+func (f *fetcher) mainLoopSendResult(j *job) {
+	f.result <- j.res
+	delete(f.jobs, j.id)
+	f.doneCnt--
+	if len(j.res.errs) != 0 {
+		return
+	}
+	// we want the first child to be the first to run not the last so
+	// add the jobs in the reverse order
+	for i := len(j.cids) - 1; i >= 0; i-- {
+		hamt := j.res.vals[string(j.cids[i].Bytes())]
+		f.mainLoopAddJob(hamt)
+	}
+	if f.idle {
+		f.launch()
+	}
+}
+
+func (f *fetcher) mainLoopLogStats() {
+	log.Infof("fetcher stats (cids, done, hits, nearMisses, misses): %d %d %d %d %d", f.cidCnt, f.doneCnt, f.hits, f.nearMisses, f.misses)
+	elapsed := time.Now().Sub(f.start).Seconds()
+	log.Infof("fetcher perf (cids/sec, jobs/sec) %f %f", float64(f.cidCnt)/elapsed, float64(f.doneCnt+f.hits+f.nearMisses+f.misses)/elapsed)
+}
+
+type batchJob struct {
+	cids []*cid.Cid
+	jobs []*job
+}
+
+func (b *batchJob) add(j *job) {
+	b.cids = append(b.cids, j.cids...)
+	b.jobs = append(b.jobs, j)
+	j.idx = -1
+}
+
+func (f *fetcher) launch() {
+	bj := batchJob{}
+
+	// always do todoFirst
+	if f.todoFirst != nil {
+		bj.add(f.todoFirst)
+		f.todoFirst = nil
+	}
+
+	// pop requets from todo list until we hit the batchSize
+	for !f.todo.empty() && len(bj.cids)+len(f.todo.top().cids) <= batchSize {
+		j := f.todo.pop()
+		bj.add(j)
+	}
+
+	if len(bj.cids) == 0 {
+		if !f.idle {
+			log.Infof("fetcher: entering idle state: no more jobs")
+		}
+		f.idle = true
+		return
+	}
+
+	// launch job
+	log.Infof("fetcher: starting batch job, size = %d", len(bj.cids))
+	f.idle = false
+	go func() {
+		ch := f.dserv.GetMany(f.ctx, bj.cids)
+		fetched := result{vals: make(map[string]*Shard)}
+		for no := range ch {
+			if no.Err != nil {
+				fetched.errs = append(fetched.errs, no.Err)
+				continue
```

Comment:
Any reason to not abort early here (close context and return)?

Reply:
Because it's unnecessary.  The fetcher can continue even if it encounters an error here.

Reply:
From what I can see, at least in https://github.com/ipfs/go-ipfs/pull/4979/files#diff-689271378656b0bd9fc790b4a0a2b784R393 we throw the result away if there are errors, so it would make sense to not fetch more stuff if we know it won't be used (I might be wrong here though)

Reply:
Note that code code also aborts the context so the amount of extra work done is minimal.